From: Phillip Lougher <phillip@lougher.org.uk>
Subject: [PATCH] corrupted cramfs filesystems cause kernel oops
Patch-mainline: Submitted Sun, 5 Nov 2006 04:02:00 +0000
References: 218237

Steve Grubb's fzfuzzer tool (http://people.redhat.com/sgrubb/files/ 
fsfuzzer-0.6.tar.gz) generates corrupt Cramfs filesystems which cause  
Cramfs to kernel oops in cramfs_uncompress_block().  The cause of the  
oops is an unchecked corrupted block length field read by  
cramfs_readpage().

This patch adds a sanity check to cramfs_readpage() which checks that  
the block length field is sensible.   The (PAGE_CACHE_SIZE << 1) size  
check is intentional, even though the uncompressed data is not going  
to be larger than PAGE_CACHE_SIZE, gzip sometimes generates  
compressed data larger than the original source data. Mkcramfs checks  
that the compressed size is always less than or equal to  
PAGE_CACHE_SIZE << 1.  Of course Cramfs could use the original  
uncompressed data in this case, but it doesn't.

Patch is against 2.6.19-rc4.

Signed-off-by:  Phillip Lougher <phillip <at> lougher.org.uk>
Acked-by: Jeff Mahoney <jeffm@suse.com>

---
 fs/cramfs/inode.c |    2 ++
 1 files changed, 2 insertions(+)

--- linux-2.6.18.orig/fs/cramfs/inode.c
+++ linux-2.6.18/fs/cramfs/inode.c
@@ -483,6 +483,8 @@ static int cramfs_readpage(struct file *
 		pgdata = kmap(page);
 		if (compr_len == 0)
 			; /* hole */
+		else if (compr_len > (PAGE_CACHE_SIZE << 1))
+			printk(KERN_ERR "cramfs: bad compressed blocksize %u\n", compr_len);
 		else {
 			mutex_lock(&read_mutex);
 			bytes_filled = cramfs_uncompress_block(pgdata,
